libexec/rtld-elf/rtld_c18n.c

```diff
@@ -0,0 +1,700 @@
+{
+	void *cur = atomic_load_explicit(&pg->cursor, memory_order_relaxed);
+	void *n_cur, *tramp;
+	do {
+		ptraddr_t n_cur_addr;
+		ptrdiff_t n_cur_len;
+
+		tramp = cheri_setbounds(cur, len);
+		n_cur_addr = roundup2(cheri_gettop(tramp), _Alignof(void *));
+		n_cur_len = cheri_gettop(cur) - n_cur_addr;
+
+		n_cur = cheri_setaddress(cur, n_cur_addr);
+		n_cur = cheri_setboundsexact(n_cur, n_cur_len);
+
+	} while (!atomic_compare_exchange_weak_explicit(
+	    &pg->cursor, &cur, n_cur,
+	    /*
+	     * Relaxed ordering is sufficient because there are no side-effects.
+	     */
+	    memory_order_relaxed, memory_order_relaxed));
+
+	assert(!cheri_gettag(tramp) || cheri_gettag(n_cur));
+	return (tramp);
+}
+
+typedef ssize_t slot_idx_t;
+
+static struct {
+	struct tramp_data *data;
+	struct tramp_map_kv {
+		_Atomic ptraddr_t key;
+		_Atomic slot_idx_t index;
+	} *map;
+	_Atomic slot_idx_t size, writers;
+	int exp;
+} tramp_table;
+
+static struct {
+	struct tramp_pg *_Atomic head;
+	atomic_flag lock;
+} tramp_pgs = {
+	.lock = ATOMIC_FLAG_INIT
+};
+
+static slot_idx_t
+tramp_table_max_load(int exp)
+{
+	/* MAX_LOAD is 37.5% of capacity. */
+	return (3 << (exp - 3));
+}
+
+static struct tramp_map_kv *
+tramp_map_new(int exp)
+{
+	struct tramp_map_kv *map;
+
+	map = xmalloc(sizeof(*map) << exp);
+	for (size_t i = 0; i < (1 << exp); ++i)
+		map[i] = (struct tramp_map_kv) {
+			.key = 0,
+			.index = -1
+		};
+	return (map);
+}
+
+static struct tramp_data *
+tramp_data_expand(int exp)
+{
+	struct tramp_data *data;
+	/*
+	 * The table only needs to be as large as the MAX_LOAD.
+	 */
+	data = realloc(tramp_table.data, sizeof(*data) * tramp_table_max_load(exp));
+	if (data == NULL)
+		rtld_fatal("realloc failed");
+	return (data);
+}
+
+/* Taken from https://github.com/skeeto/hash-prospector/issues/19 */
+static uint32_t
+pointerHash(uint64_t key) {
+	uint32_t x = key ^ (key >> 32);
+	x ^= x >> 16;
+	x *= 0x21f0aaadU;
+	x ^= x >> 15;
+	x *= 0xd35a2d97U;
+	x ^= x >> 15;
+	return (x);
+}
+
+static slot_idx_t
+nextSlot(slot_idx_t slot, uint32_t hash, int exp)
+{
+	uint32_t mask = (1 << exp) - 1;
+	uint32_t step = (hash >> (32 - exp)) | 1;
+	return ((slot + step) & mask);
+}
+
+static void
+resizeTable(int exp)
+{
+	struct tramp_map_kv *map;
+	struct tramp_data *data;
+	ptraddr_t key;
+	uint32_t hash;
+	slot_idx_t size, slot;
+
+	assert(0 < exp && exp < 32);
+
+	free(tramp_table.map);
+	data = tramp_data_expand(exp);
+	map = tramp_map_new(exp);
+	size = atomic_load_explicit(&tramp_table.size, memory_order_relaxed);
+
+	for (slot_idx_t idx = 0; idx < size; ++idx) {
+		key = (ptraddr_t)data[idx].target;
+		hash = pointerHash(key);
+		slot = hash;
+		do
+			slot = nextSlot(slot, hash, exp);
+		while (atomic_load_explicit(
+		    &map[slot].key, memory_order_relaxed) != 0);
+		atomic_store_explicit(&map[slot].key, key,
+		    memory_order_relaxed);
+		atomic_store_explicit(&map[slot].index, idx,
+		    memory_order_relaxed);
```

Comment:
Does it matter there's a period where you have a valid key but no index?

Reply:
`resizeTable` is only called when we have exclusive access to the table, so this doesn't matter.

Reply:
Then why all the atomics? If it's under a lock you don't need them, and shouldn't have them because:
(a) it's potentially slower (though not really for relaxed)
(b) it's confusing, because it strongly suggests there *is* concurrent access here to be careful about, yet the code has bugs if there is concurrent access because the atomics aren't sufficient

Reply:
I think if you use `=` to assign to a C atomic variable then the generated code will do a store-release, which is why I'm doing the relaxed store explicitly. But I'll add a comment about that.

Reply:
I know. That's not what I'm asking. I'm asking why they're atomics in the first place.

Reply:
And if it has exclusive access, why is it spinning on reading the key?

Reply:
The atomics are needed because the table is concurrently accessed during normal insertions and reads.

And that is actually a do-while loop without curly braces, which can be confusing...

Reply:
I don't understand what your model is then. Which of the following is it:
1. Multiple readers or a single writer, never both
2. Multiple readers and a single writer, sometimes both
If the former there is no need for atomics beyond your reader/writer lock. If the latter I'm concerned about readers seeing intermediate state whilst the writer inserts.

Reply:
Neither actually. The model is multiple readers and multiple writers, except when resizing when there's only one writer.